jwql/instrument_monitors/common_monitors/bad_pixel_monitor.py

```diff
@@ -0,0 +1,1066 @@
+#! /usr/bin/env python
+
+"""This module contains code for the bad/dead pixel monitor, XXX
+XXXXXXXXXXXXXX THIS DOCSTRING NEEDS TO BE UPDATED
+
+If enough new files for a given instrument/aperture combination
+(currently the files must be identified as dark current files in the
+``exp_type`` header keyword) are present in the filesystem at the time
+the ``dark_monitor`` is called, the files are first run through the the
+appropriate pipeline steps to produce slope images.
+
+A mean slope image as well as a standard deviation slope image is
+created by sigma-clipping on a pixel by pixel basis. The mean and
+standard deviation images are saved to a fits file, the name of which
+is entered into the ``<Instrument>DarkCurrent`` database table.
+
+The mean slope image is then normalized by an existing baseline slope
+image. New hot pixels are identified as those with normalized signal
+rates above a ``hot_threshold`` value. Similarly, pixels with
+normalized signal rates below a ``dead_threshold`` are flagged as new
+dead pixels.
+
+The standard deviation slope image is normalized by a baseline
+(historical) standard deviation image. Pixels with normalized values
+above a noise threshold are flagged as newly noisy pixels.
+
+New hot, dead, and noisy pixels are saved to the ``DarkPixelStats``
+database table.
+
+Next, the dark current in the mean slope image is examined. A histogram
+of the slope values is created for the pixels in each amplifier, as
+well as for all pixels on the detector. In all cases, a Gaussian is fit
+to the histogram. Currently for NIRCam and NIRISS, a double Gaussian is
+also fit to the histogram from the entire detector.
+
+The histogram itself as well as the best-fit Gaussian and double
+Gaussian parameters are saved to the DarkDarkCurrent database table.
+
+
+Author
+------
+
+    - Bryan Hilbert
+
+Use
+---
+
+    This module can be used from the command line as such:
+
+    ::
+
+        python bad_pixel_monitor.py
+
+
+
+Development notes:
+
+This can't be used with NIRCam since NIRCam has no internal lamp and therefore
+will not be taking any more internal flat field images. Could perhaps be used with
+a series of external undithered observations...
+
+Dead pixel algorithem was designed to use flats for all instruments except MIRI, which uses
+darks. (Would anything useful result if using darks for NIRCam?)
+
+Templates to use: FGS_INTFLAT, NIS_LAMP, NRS_LAMP, MIR_DARK
+
+
+
+
+
+
+
+
+
+
+
+"""
+
+from copy import deepcopy
+import datetime
+import logging
+import os
+
+from astropy.io import ascii, fits
+from astropy.time import Time
+from jwst.datamodels import dqflags
+from jwst_reffiles.bad_pixel_mask import bad_pixel_mask
```

Comment:
I think this is a new dependency for the project.  Could you add it appropriately to `setup.py`, `requirements.txt`, and `environment-3.6.yml`?

Reply:
Done